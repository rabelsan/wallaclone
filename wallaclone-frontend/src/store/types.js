export const AUTH_LOGIN_REQUEST = 'auth/LOGIN_REQUEST';
export const AUTH_LOGIN_SUCCESS = 'auth/LOGIN_SUCCESS';
export const AUTH_LOGIN_FAILURE = 'auth/LOGIN_FAILURE';

export const AUTH_LOGOUT = 'auth/LOGOUT';

export const ADS_REQUEST = 'ads/REQUEST';
export const ADS_SUCCESS = 'ads/SUCCESS';
export const ADS_FAILURE = 'ads/FAILURE';

export const AD_REQUEST = 'ad/REQUEST';
export const AD_REQUEST_SUCCESS = 'ad/REQUEST_SUCCESS';
export const AD_REQUEST_FAILURE = 'ad/REQUEST_FAILURE';
export const AD_DELETE_REQUEST = 'ad/DELETE_REQUEST';
export const AD_DELETE_SUCCESS = 'ad/DELETE_SUCCESS';
export const AD_DELETE_FAILURE = 'ad/DELETE_FAILURE';
export const AD_CREATE_REQUEST = 'ad/CREATE_REQUEST';
export const AD_CREATE_SUCCESS = 'ad/CREATE_SUCCESS';
export const AD_CREATE_FAILURE = 'ad/CREATE_FAILURE';
export const AD_RESET_DETAILS = 'ad/RESET_DETAILS';

export const TAGS_REQUEST = 'tags/REQUEST';
export const TAGS_SUCCESS = 'tags/SUCCESS';
export const TAGS_FAILURE = 'tags/FAILURE';

