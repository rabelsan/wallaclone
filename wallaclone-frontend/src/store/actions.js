import {
  AUTH_LOGIN_REQUEST,
  AUTH_LOGIN_FAILURE,
  AUTH_LOGIN_SUCCESS,
  AUTH_LOGOUT,
  ADS_REQUEST,
  ADS_SUCCESS,
  ADS_FAILURE,
  AD_REQUEST,
  AD_REQUEST_SUCCESS,
  AD_REQUEST_FAILURE,
  AD_DELETE_REQUEST,
  AD_DELETE_SUCCESS,
  AD_DELETE_FAILURE,
  AD_CREATE_REQUEST,
  AD_CREATE_SUCCESS,
  AD_CREATE_FAILURE,
  AD_RESET_DETAILS,
  TAGS_REQUEST,
  TAGS_SUCCESS,
  TAGS_FAILURE,
} from './types';

import * as adsApi from '../api/adverts';

export const authLoginRequest = () => ({
  type: AUTH_LOGIN_REQUEST,
});
  
export const authLoginFailure = error => ({
  type: AUTH_LOGIN_FAILURE,
  error: true,
  payload: error,
});
  
export const authLoginSuccess = loggedUserToken => ({
  type: AUTH_LOGIN_SUCCESS,
  payload: loggedUserToken,
});

export const login = credentials => {
    return async function (dispatch, getState, { history, api }) {
      dispatch(authLoginRequest());
      try {
        const loggedUserToken =  await api.auth.login(credentials);
        dispatch(authLoginSuccess(loggedUserToken));
        history.push('/adverts');
      } catch (error) {
        dispatch(authLoginFailure(error));
      }
    };
  };

  export const adsRequest = () => ({
    type: ADS_REQUEST,
  });

  export const adsFailure = error => ({
    type: ADS_FAILURE,
    payload: error,
  });

  export const adsSuccess = adsList => ({
    type: ADS_SUCCESS,
    payload: adsList,
  });

  export const adRequest = () => ({
    type: AD_REQUEST,
  });

  export const adRequestFailure = error => ({
    type: AD_REQUEST_FAILURE,
    payload: error,
  });

  export const adRequestSuccess = (ad, isNew) => ({
    type: AD_REQUEST_SUCCESS,
    payload: { ad, isNew }
  });

  export const adCreateRequest = adsList => ({
    type: AD_CREATE_REQUEST,
  });

  export const adCreateSuccess = (result) => ({
    type: AD_CREATE_SUCCESS,
    payload: result,
  });

  export const adCreateFailure = error => ({
    type: AD_CREATE_FAILURE,
    payload: error,
  });

  export const adDeleteRequest = adsList => ({
    type: AD_DELETE_REQUEST,
  });

  export const adDeleteSuccess = (result) => ({
    type: AD_DELETE_SUCCESS,
    payload: result,
  });

  export const adDeleteFailure = error => ({
    type: AD_DELETE_FAILURE,
    payload: error,
  });

  export const adResetDetails = () => ({
    type: AD_RESET_DETAILS,
  });

  export const createAd = id => {
    return async function (dispatch, getState, { history, api }) {
      dispatch(adCreateRequest());
      try {
        const ad = await api.adverts.createAdvert(id);
        dispatch(adCreateSuccess(ad.result));
      } catch (error) {
        dispatch(adCreateFailure(error));
      }
    };
  };

  export const deleteAd = id => {
    return async function (dispatch, getState, { history, api }) {
      dispatch(adDeleteRequest());
      try {
        const result = await api.adverts.deleteAdvert(id);
        dispatch(adDeleteSuccess(result));
      } catch (error) {
        dispatch(adDeleteFailure(error));
      }
    };
  };

  export const loadAd = (id, isNew = false) => {
    return async function (dispatch, getState, { history, api }) {
      dispatch(adRequest());
      try {
        const ad =  await api.adverts.getAdvert(id);
        dispatch(adRequestSuccess(ad.result, isNew));
      } catch (error) {
        dispatch(adRequestFailure(error));
      }
    };
  };

  export const loadAds = filters => {
    return async function (dispatch, getState) {
      dispatch(adsRequest());
      try {
        const ads =  await adsApi.getAdverts(filters);
        dispatch(adsSuccess(ads.result.rows));
      } catch (error) {
        dispatch(adsFailure(error));
      }
    };
  };

  export const tagsRequest = () => ({
    type: TAGS_REQUEST,
  });

  export const tagsFailure = error => ({
    type: TAGS_FAILURE,
    payload: error,
  });

  export const tagsSuccess = tagsList => ({
    type: TAGS_SUCCESS,
    payload: tagsList,
  });

  export const loadTags = () => {
    return async function (dispatch, getState, { history, api }) {
      dispatch(tagsRequest());
      try {
        const tags =  await api.adverts.getTags();
        dispatch(tagsSuccess(tags.result));
      } catch (error) {
        dispatch(tagsFailure(error));
      }
    };
  };
    
  export const authLogout = () => {
    return {
      type: AUTH_LOGOUT,
    };
  };
   