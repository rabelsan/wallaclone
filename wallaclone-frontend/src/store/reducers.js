import * as types from './types';

const initialState = {
  auth: null,
  ads: {
    adverts: [],
    loading: false,
    error: null,
  },
  adDetails: {
    processing: false,
    errorChange: null,
    advert: null,
    isNew: false,
  },
  tags: {
    list: [],
    error: null,
  },
  ui: {
    loading: false,
    error: null,
  },
};

export const auth = (state = initialState.auth, action) => {
  switch (action.type) {
    case types.AUTH_LOGIN_SUCCESS:
      // login
      return action.payload;
    case types.AUTH_LOGOUT:
      // logout
      return null;
    default:
      return state;
  }
};

export const ads = (state = initialState.ads, action) => {
  switch (action.type) {
    case types.ADS_REQUEST:
      return { ...state, adverts: [], loading:true, error: null };
    case types.ADS_SUCCESS:
      return { ...state, adverts: action.payload, loading: false, error: null };
    case types.ADS_FAILURE:
      return { ...state, adverts: [], loading: false, error: action.payload };
    default:
      return state;
  }
};

export const adDetails = (state = initialState.adDetails, action) => {
  switch (action.type) {
    case types.AD_REQUEST:
      return { ...state, processing: true, errChange: null, advert: null, isNew: false };
    case types.AD_REQUEST_SUCCESS:
      return { ...state, processing: false, errChange: null, advert: action.payload.ad, isNew: action.payload.isNew };
    case types.AD_REQUEST_FAILURE:
        return { ...state, processing: false, errChange: action.payload, advert:null, isNew: false };
    case types.AD_DELETE_REQUEST:
      return { ...state, processing: true, errChange: null, advert: action.payload, isNew: false };
    case types.AD_DELETE_SUCCESS:
      return { ...state, processing: false, errChange: null, advert: null, isNew: false };
    case types.AD_DELETE_FAILURE:
        return { ...state, processing: false, errChange: action.payload, advert:null, isNew: false };
    case types.AD_CREATE_REQUEST:
      return { ...state, processing: true, errChange: null, advert: null, isNew: false };
    case types.AD_CREATE_SUCCESS:
      return { ...state, processing: false, errChange: null, advert: action.payload, isNew: true };
    case types.AD_CREATE_FAILURE:
      return { ...state, processing: false, errChange: action.payload, advert: null, isNew: false }; 
    case types.AD_RESET_DETAILS:
      return { ...state, processing: false, errChange: null, advert: null, isNew: false };  
    default:
      return state;
  }
};

export const tags = (state = initialState.tags, action) => {
  switch (action.type) {
    case types.TAGS_REQUEST:
      return { ...state, list: [], error: null };
    case types.TAGS_SUCCESS:
      return { ...state, list: action.payload, error: null };
    case types.TAGS_FAILURE:
      return { ...state, list: [], error: action.payload };
    default:
      return state;
  }
};

export const ui = (state = initialState.ui, action) => {
    if (action.error) {
      return { ...state, error: action.payload, loading: false };
    }
    switch (action.type) {
      case types.AUTH_LOGIN_REQUEST:
        return { ...state, error: null, loading: true };
      case types.AUTH_LOGIN_SUCCESS:
        return { ...state, error: null, loading: false };
      default:
        return state;
    }
  };