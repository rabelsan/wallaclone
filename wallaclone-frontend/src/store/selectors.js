export const getLoggedUserToken = state => state.auth;

export const getUi = state => state.ui;

export const getAds = state => state.ads;

export const getAdDetails = state => state.adDetails; 

export const getAdvertById = id => state => {
  let result = getAds(state);
  if (!result) {
    return null;
  }
  return result.adverts.find(ad => ad._id === id); 
};

export const getTags = state => state.tags;
