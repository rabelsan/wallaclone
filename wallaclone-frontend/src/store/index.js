import { createStore, combineReducers, applyMiddleware } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import { connectRouter, routerMiddleware } from 'connected-react-router';

import thunk from 'redux-thunk';
import logger from 'redux-logger';

import * as reducers from './reducers';
import * as api from '../api';

export function configureStore(preloadedState, { history }) {
  const reducer = combineReducers({
    router: connectRouter(history),
    ...reducers,
  });
  const middlewares = [
    routerMiddleware(history),
    thunk.withExtraArgument({ history, api }),
    logger,
  ];
  const store = createStore(
    reducer,
    preloadedState,
    composeWithDevTools(applyMiddleware(...middlewares))
    // window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
  );
  return store;
}
