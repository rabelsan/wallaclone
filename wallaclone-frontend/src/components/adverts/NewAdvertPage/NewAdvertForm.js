import React, {useState} from 'react';
import { connect } from 'react-redux';

import T from 'prop-types';
import { Button, Radio, Input, InputNumber, Row, Col } from 'antd';

import { getTags } from '../../../store/selectors';

import TagsSelect from '../TagsSelect';
import { FormField, InputImage } from '../../shared';
import * as numbers from '../../../utils/numbers';

import { saleOptions, MIN_PRICE, MAX_PRICE } from '../definitions';

import styles from './NewAdvertForm.module.css';

function NewAdvertForm ({ onSubmit, list }) {
  const [state, setState] = useState({
    name: '',
    price: 0,
    tags: [],
    photo: null,
    sale: saleOptions.sell.value,
  });

  const canSubmit = () => {
    const { name, price, tags } = state;

    return [
      // valid name
      !!name,
      // valid price
      !Number.isNaN(price) && Number.isFinite(price) && price >= 0,
      // valid tags
      !!tags.length,
    ].every(validation => validation); // all validations pass
  };

  const getFormData = () => {
    const { name, price, tags, sale, photo } = state;
    const formData = new FormData();
    formData.append('name', name);
    formData.append('sale', sale === saleOptions.sell.value);
    formData.append('price', price);
    tags.forEach((tag, index) => formData.append(`tags[${index}]`, tag));
    if (photo) formData.append('photo', photo);
    return formData;
  };

  const handleNameChange = ev => setState({ ...state, name: ev.target.value });
  const handlePriceChange = price => setState({ ...state, price });
  const handleTagsChange = tags => setState({ ...state, tags });
  const handlePhotoChange = photo => setState({ ...state, photo });
  const handleSaleChange = ev => setState({ ...state, sale: ev.target.value });

  const handleSubmit = ev => {
    ev.preventDefault();
    onSubmit(getFormData());
  };

  const { name, price, tags, sale } = state;
    
  return (
    <form onSubmit={handleSubmit}>
      <Row className={styles.form}>
        <Col span={11}>
          <FormField label="Name">
            <Input
              placeholder="Name"
              onChange={handleNameChange}
              value={name}
            />
          </FormField>
          <FormField label="Price">
            <InputNumber
              {...numbers}
              className={styles.price}
              min={MIN_PRICE}
              max={MAX_PRICE}
              onChange={handlePriceChange}
              value={price}
            />
          </FormField>
        </Col>
        <Col span={11} offset={2}>
          <FormField label="Tags">
            <TagsSelect onChange={handleTagsChange} options={list} value={tags}/>
          </FormField>
          <FormField label="Type">
            <Radio.Group
              options={[saleOptions.sell, saleOptions.buy]}
              onChange={handleSaleChange}
              value={sale}
            />
          </FormField>
        </Col>
        <Col span={24}>
          <FormField label="Photo">
            <InputImage type="file" onChange={handlePhotoChange} />
          </FormField>
          <Button
            className={styles.button}
            type="primary"
            htmlType="submit"
            disabled={!canSubmit()}
            block
          >
            Up!
          </Button>
        </Col>
      </Row>
    </form>
  );
}

NewAdvertForm.propTypes = {
  onSubmit: T.func.isRequired,
  list: T.arrayOf(T.string),
};

export default connect(getTags) (NewAdvertForm);
