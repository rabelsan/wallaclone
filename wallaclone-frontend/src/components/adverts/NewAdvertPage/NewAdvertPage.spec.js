import React from 'react';
import { shallow } from 'enzyme';

import NewAdvertForm from './NewAdvertForm';
import { createAd }  from '../../../store/actions';
import Layout from '../../layout';
import { Alert, Divider } from 'antd';

describe('NewAdvertPage', () => {
  const newAd = createAd;
  const errChange = 'Error';

  const handleSubmit = advertFormData => {
    newAd(advertFormData);
  };

  const render = () => shallow(
    <Layout title="New advert">
      <Divider>Create an advert</Divider>
      <NewAdvertForm onSubmit={handleSubmit} />
      {errChange && (
        <Alert
          closable
          message={errChange}
          showIcon
          type="error"
        />
      )}
    </Layout>
  );

  
  test('snapshot', () => {
    const wrapper = render();
    expect(wrapper).toMatchSnapshot();
  });
  test('snapshot with title', () => {
    const wrapper = render();
    expect(wrapper.find('Title').getElement().props.children).toBe('New advert');
    expect(wrapper).toMatchSnapshot();
  });

  test('NewAdvertForm should receive onSubmit', () => {
    const wrapper = render();
    expect(wrapper.find('Connect(NewAdvertForm)').getElement().props.onSubmit).toBeDefined();
  });
});
