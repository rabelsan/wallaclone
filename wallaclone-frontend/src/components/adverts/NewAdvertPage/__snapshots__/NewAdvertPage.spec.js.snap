// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewAdvertPage snapshot 1`] = `
<Layout
  className="layout"
>
  <Header />
  <Content
    className="main"
  >
    <Row
      className="row"
    >
      <Col
        className="col"
        offset={4}
        span={16}
      >
        <Title
          level={2}
        >
          New advert
        </Title>
        <div>
          <Divider>
            Create an advert
          </Divider>
          <Connect(NewAdvertForm)
            onSubmit={[Function]}
          />
          <Alert
            closable={true}
            message="Error"
            showIcon={true}
            type="error"
          />
        </div>
      </Col>
    </Row>
  </Content>
  <Footer
    className="footer"
  >
    © 2021 Ramón Beltrán
  </Footer>
</Layout>
`;

exports[`NewAdvertPage snapshot with title 1`] = `
<Layout
  className="layout"
>
  <Header />
  <Content
    className="main"
  >
    <Row
      className="row"
    >
      <Col
        className="col"
        offset={4}
        span={16}
      >
        <Title
          level={2}
        >
          New advert
        </Title>
        <div>
          <Divider>
            Create an advert
          </Divider>
          <Connect(NewAdvertForm)
            onSubmit={[Function]}
          />
          <Alert
            closable={true}
            message="Error"
            showIcon={true}
            type="error"
          />
        </div>
      </Col>
    </Row>
  </Content>
  <Footer
    className="footer"
  >
    © 2021 Ramón Beltrán
  </Footer>
</Layout>
`;
